man/cat1/moments.1



MOMENTS(1)               TIPSY COMMANDS                MOMENTS(1)



NAME
     moments - write the shape as a funtion of radius

SYNOPSIS
     moments _b_o_x-_n_u_m_b_e_r _c_e_n_t_e_r-_b_o_x-_n_u_m_b_e_r _p_a_r_t_i_c_l_e  _b_i_n_n_i_n_g  _b_i_n-
     _n_u_m_b_e_r _f_i_l_e_n_a_m_e _m_i_n_i_m_u_m__r_a_d_i_u_s

DESCRIPTION
     moments is a command that produces a file of name, _f_i_l_e_n_a_m_e,
     that contains various physical parameters that relate to the
     shape and orientation of the density distribution for parti-
     cles  of type _p_a_r_t_i_c_l_e in box, _b_o_x-_n_u_m_b_e_r. If the boxes _b_o_x-
     _n_u_m_b_e_r or _c_e_n_t_e_r-_b_o_x-_n_u_m_b_e_r are not loaded an error is  gen-
     erated.   The  shapes  and orientations are calculated using
     the iterative technique described in  Katz  (Ap.J.  368:325,
     1991).  The initial guess for the center of the density dis-
     tribution is taken to be the center of mass of  box  _c_e_n_t_e_r-
     _b_o_x-_n_u_m_b_e_r. The _b_i_n_n_i_n_g argument refers to how the particles
     are binned, either linearly or log  in  radius.   There  are
     _b_i_n-_n_u_m_b_e_r  equal size bins distributed equally in radius or
     log radius between _m_i_n_i_m_u_m__r_a_d_i_u_s and the size of  box  _b_o_x-
     _n_u_m_b_e_r  depending  on the _b_i_n_n_i_n_g argument.  The output file
     has the following columns: radius,  b/a,  c/a,  phi,  theta,
     psi,  center-x,center-y,and center-z where b/a refers to the
     ratio between the intermediate axis and the  long  axis  and
     c/a  refers  to  the  ratio between the short and long axes.
     Phi, theta and psi refer to the Euler angles  that  describe
     the orientation of the density distribution.

OPTIONS
  Particle
     The possible choices for _p_a_r_t_i_c_l_e are:

          all
             Use the all the particles in box _b_o_x-_n_u_m_b_e_r

          dark
             Use only the dark matter particles in box _b_o_x-_n_u_m_b_e_r

          gas
             Use only the gas particles in box _b_o_x-_n_u_m_b_e_r

          star
             Use only the star particles in box _b_o_x-_n_u_m_b_e_r

  Binning
     The possible choices for _b_i_n_n_i_n_g are:

          lin
             Make  _b_i_n-_n_u_m_b_e_r  bins  spaced  equally  in   radius
             between  _m_i_n_i_m_u_m-_r_a_d_i_u_s  and  the  size  of box _b_o_x-
             _n_u_m_b_e_r



KQ Release 2.0     Last change: 22 MARCH 1994                   1






MOMENTS(1)               TIPSY COMMANDS                MOMENTS(1)



          log
             Make _b_i_n-_n_u_m_b_e_r bins spaced equally  in  log  radius
             between  _m_i_n_i_m_u_m-_r_a_d_i_u_s  and  the  size  of box _b_o_x-
             _n_u_m_b_e_r

SEE ALSO
     profile(1), rotationcurve(1),

BUGS
     The technique doesn't always  converge,  particularily  when
     the  density  distribution  is very clumpy.  Things converge
     much better when the box _c_e_n_t_e_r-_b_o_x-_n_u_m_b_e_r contains the true
     center for the density distribution in box _b_o_x-_n_u_m_b_e_r and is
     as small as possible.









































KQ Release 2.0     Last change: 22 MARCH 1994                   2





code/ChangeLog
Wed Jan 11 13:09:36 1995  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (SRCS): added alloca.c so it will be included in the
	distribution.

Wed Dec 28 13:42:47 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* readmark.c: close file after it is read.

	* main.c: RedrawViewport(): make can_size the maximum of
	can_height and can_width.

	* canvas_resize_proc.c: Same as above.

Mon Dec  5 09:39:44 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* xray_lum_load.c, star_lum.c, star_lum_redshift.c: Changed
	argument to my_gets() to blank string.

Fri Dec  2 08:21:50 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (READLINE_SRC): delete config.h.
	(tipsy): added dependence on ALLOCA.

	* fits.c: Added test for endianness.

Thu Dec  1 15:14:58 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* vista.c: fixed initialization of quantity[].

	* find_shape.c: Forced inertia_cm to be symmetric.  This avoid
	problems with jacobi().

Tue Nov  1 11:26:56 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in : fits_xray.c is now gone.  fits() is used instead.
	defs.h, xray.c, vista.c, fits.c and arguments.c are also changed.

Mon Oct 31 09:43:29 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* main.c: RedrawColorbar(): fixed off-by-one drawing errors.

Wed Oct 12 14:27:16 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Created version 2.1.2.

	* absorb.c: New function for calculating aborbtion spectra.

Tue Sep  6 16:06:02 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline/: Replaced with readline-2.0 from prep.ai.mit.edu.

	* my_gets.c: Changed cast to match new external from readline/.

Fri Jul 29 10:25:49 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Ensure that the axes projections do not overflow.

Fri Jul 15 17:28:15 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* profile.c: Check to see if fopen() fails.

Wed Jul 13 09:46:18 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h: Decreased INTMAX to 2^29-1.  (One bit for sign, one bit
	to do INTMAX - (-INTMAX), and one bit for zoom out.

Tue Jul 12 15:55:17 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* canvas_resize_proc.c: Call plot_sub() and draw_label() for
	servers that automatically do backing store and don't generate an
	expose event.

	* main.c: ZoomIn(), ZoomOut(), Pan(), SetBox(): Ignore button
	presses in inactive windows.
	
	* main.c, window.c: Changed ResizeRequest to ConfigureNotify.

Sat Jun  4 13:12:06 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* fits.c, and fits_xray.c: Changed write() to fwrite() for
	portability.

Thu May  5 10:53:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Increase clipping range of velocities to -2*INTMAX <
	v < 2*INTMAX.
	* defs.h: Change INTMAX to 2^30-1 to accommodate the above
	clipping.

Tue Apr 19 18:00:14 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* comm.h: NSK: added pcenter and writeboxrot commands.

Mon Apr 18 14:14:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Add "unzoom" button.

Fri Apr 15 16:01:56 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Fixed bug with color of marked particles.
	Deleted extra conversion to screen coordinates.
	
	* maxpos.c: Make the margins a little smaller.

Thu Mar 31 14:03:11 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Changed XDrawArc to XFillArc for vector draws.

Fri Mar  4 18:09:25 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Added Neal's changes for read_vector() and "master".

Mon Feb 28 11:51:39 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* openascii.c: Added code from Peter Teuben to uncompress files.

Thu Feb 24 13:21:50 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Changed current_plot logical to current_project and
	current_color.

Tue Feb 22 11:24:05 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline: Updated to readline from gdb-1.12

Thu Feb 17 11:46:07 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Implimented a software "z-buffer".  Only draw
	particles with the highest particle color.

Wed Feb 16 13:21:35 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Merged plot_all and plot_vel into one routine.  A
	new routine all_color() is created.  Changes were also needed in
	the view_*() routines.

Sat Nov 13 11:35:01 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Fixed typo for increasing ndark.

Wed Nov  3 14:35:56 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Deleted skip of XClearWindow on first time through.

Thu Oct 28 10:23:50 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* hard.c: Added movie parameter.

Tue Oct 26 11:29:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* alloca.c: New file in case system doesn't provide alloca().

	* vista.c: Included <stdlib.h> for malloc() definition.

Mon Oct 25 19:16:52 1993  Tom Quinn  (trq@virgo)

	* main.c: Attempt to find a depth 8 PseudoColor visual.

Thu Oct 21 08:56:32 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Set backgrounds of gc_color[] to black.

Tue Oct 19 10:56:36 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h, main.h, make_macro.c, read_macro.c, write_macro.c
	run_macro.c: re-vamped macro storage using linked lists.

Mon Oct 18 15:11:34 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* draworient.c: Added cast from Arg to long.

Fri Oct 15 10:04:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Added a check for index < 1.

	* plot_sub.c: Added a check for binary_loaded.
	Added check for boxes_loaded[0].

Wed Oct  6 09:13:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* zoom_window.c: Added zoom command.  Also changed comm.h and
	arguments.c.

Tue Oct  5 17:27:43 1993  Tom Quinn  (trq@hermes.astro.washington.edu

code/moments.c
#include "defs.h"
#include "fdefs.h"
void
moments(job)
    char job[MAXCOMM] ;
{
    char command[MAXCOMM] ;
    char particle_type[MAXCOMM] ;
    char bin_type[MAXCOMM] ;
    int box ;
    int center_box ;
    int number_bins ;
    double min_radius ;
    Real center[MAXDIM] ;
    double bin_size ;
    int i ;

    if (sscanf(job,"%s %d %d %s %s %d %s %lf",command,&box,&center_box,
	    particle_type,bin_type,&number_bins,
	    hardfile.name,&min_radius) == 8) { 
	if (boxes_loaded[box] && boxes_loaded[center_box] &&
		number_bins <= MAXBIN) {
	    if (strcmp(bin_type,"lin") == 0) {
		bin_size = (boxes[box].size / 2.0  - min_radius) /
			(double)number_bins ;
	    }
	    else if (strcmp(bin_type,"log") == 0) {
		min_radius = log10(min_radius) ;
		bin_size = (log10(boxes[box].size / 2.0 ) - min_radius) /
			(double)number_bins ;
	    }
	    else {
		printf("<sorry, %s not a bin type, %s>\n",bin_type,title) ;
		return ;
	    }
	    if (strcmp(particle_type,"dark") != 0 && 
		    strcmp(particle_type,"gas") != 0 &&
		    strcmp(particle_type,"star") != 0 &&
		    strcmp(particle_type,"all") !=0) {
		printf("<sorry, %s is not a particle type, %s>\n",
			particle_type,title) ;
		return ;
	    }
	    hardfile.ptr = fopen(hardfile.name,"w") ;
	    for (i = 1 ;i <= number_bins ;i++) {
		if (strcmp(particle_type,"dark") == 0){
		    setvec(center,boxes[center_box].dark_com) ;
		}
		else if (strcmp(particle_type,"star") == 0){
		    setvec(center,boxes[center_box].star_com) ;
		}
		else if (strcmp(particle_type,"gas") == 0){
		    setvec(center,boxes[center_box].gas_com) ;
		}
		else{
		    setvec(center,boxes[center_box].total_com) ;
		}
		fit_radius = bin_size * (double)(i) ;
		if (strcmp(bin_type,"lin") == 0) {
		    fit_radius += min_radius ;
		}
		else if (strcmp(bin_type,"log") == 0) {
		    fit_radius = pow(10.,fit_radius + min_radius) ;
		}
		find_shape(particle_type,box,center) ;
		fprintf(hardfile.ptr,"%g %g %g ",fit_radius,ba,ca) ;
		fprintf(hardfile.ptr,"%g %g %g ",phi,theta,psi) ;
		fprintf(hardfile.ptr,"%g %g %g\n",center[0],
			center[1],center[2]) ;
	    }
	    fclose(hardfile.ptr) ;
	}
	else if (!boxes_loaded[box]) {
	    printf("<sorry, box %d is not loaded, %s>\n",box,title) ;
	}
	else if (!boxes_loaded[center_box]) {
	    printf("<sorry, box %d is not loaded, %s>\n",center_box,title) ;
	}
	else {
	    printf("<%d is too many bins, %s>\n",number_bins,title) ;
	}
    }
    else {
	input_error(command) ;
    }
}
